client/session.go

```diff
@@ -1625,74 +1661,163 @@ func (s *session) FetchBlocksFromPeers(
-	for _, peer := range peers {
+	for idx, peer := range peers.peers {
+		idx := idx
 		peer := peer
 
 		wg.Add(1)
 		go func() {
-			defer wg.Done()
-			var err error
-
-			switch version {
-			case FetchBlocksMetadataEndpointV1:
-				err = s.streamBlocksMetadataFromPeer(namespace, shard,
-					peer, start, end, metadataCh, progress)
-			case FetchBlocksMetadataEndpointV2:
-				err = s.streamBlocksMetadataFromPeerV2(namespace, shard,
-					peer, start, end, metadataCh, resultOpts, progress)
-			// Should never happen - we validate the version before this function is
-			// ever called
-			default:
-				err = errInvalidFetchBlocksMetadataVersion
+			defer func() {
+				// Success or error counts towards a response
+				atomic.AddInt32(&responded, 1)
+
+				// Decrement pending
+				progress.metadataFetches.Update(float64(atomic.AddInt64(&pending, -1)))
+
+				// Mark done
+				wg.Done()
+			}()
+
+			var currPageToken pageToken
+			condition := func() bool {
+				level := consistencyLevel
+				majority := int(majority)
+				enqueued := int(enqueued)
+				success := int(atomic.LoadInt32(&success))
+				return !s.readConsistencyAchieved(level, majority, enqueued, success) &&
+					getAbortError() == nil
 			}
+			for condition() {
+				var err error
+				switch version {
+				case FetchBlocksMetadataEndpointV1:
+					currPageToken, err = s.streamBlocksMetadataFromPeer(namespace, shardID,
+						peer, start, end, currPageToken, metadataCh, progress)
+				case FetchBlocksMetadataEndpointV2:
+					currPageToken, err = s.streamBlocksMetadataFromPeerV2(namespace, shardID,
+						peer, start, end, currPageToken, metadataCh, resultOpts, progress)
+				default:
+					// Should never happen - we validate the version before this function is
+					// ever called
+					err = xerrors.NewNonRetryableError(errInvalidFetchBlocksMetadataVersion)
+					setError(idx, err)
+					return
+				}
 
-			if err != nil {
-				errLock.Lock()
-				defer errLock.Unlock()
-				errLen++
-				multiErr = multiErr.Add(err)
+				// Set error or success if err is nil
+				setError(idx, err)
+
+				// Check exit criteria
+				if err != nil && xerrors.IsNonRetryableError(err) {
+					setAbortError(err)
+					return // Cannot recover from this error, so we break from the loop
+				}
+				if err == nil {
+					atomic.AddInt32(&success, 1)
+					return
+				}
 			}
-			progress.metadataFetches.Update(float64(atomic.AddInt64(&pending, -1)))
 		}()
 	}
 
 	wg.Wait()
 
-	if errLen == len(peers) {
-		return multiErr.FinalError()
+	if err := getAbortError(); err != nil {
+		return err
 	}
-	return nil
+
+	errs := getErrors()
+	return s.readConsistencyResult(consistencyLevel, majority, enqueued,
+		atomic.LoadInt32(&responded), int32(len(errs)), errs)
 }
 
+// pageToken is just an opaque type that needs to be downcasted to expected
+// page token type, this makes it easy to use the page token across the two
+// versions
+type pageToken interface{}
```

Comment:
Could you add a TODO to delete this along with the V1 endpoint like the other comments we have

Reply:
Sure thing.